File: src/api/token.js
const bcrypt = require('bcryptjs');
const bodyParser = require('body-parser');
const jsonParser = bodyParser.json();

const createjwt = require("../util/createjwt");
const logger = require("../logger");
const app = require("../app");

const user_dao = require("../dao/user");
const access_token_dao = require("../dao/access_token");

// data='{"name": "xyz", "password": "abc"}'
// curl -X POST 127.0.0.1:8080/api/token --data "${data}" -H "Content-Type: application/json"
app.post('/api/token', jsonParser, async function(req, res){
	logger.debug("POST /api/token", req.body.name);

	const user = await user_dao.get_by_name(req.body.name);
	if (!user) {
		return res.status(404).json({
			message: "User not found"
		});
	}

	if (req.body.password){
		// login with username/password (from web-ui)
		const success = bcrypt.compareSync(req.body.password, user.hash);
		if (!success) {
			return res.status(401).json({
				message: "Invalid password"
			});
		}
		const token = createjwt(user, {
			audience: "management"
		});
		res.send(token);

	} else if (req.body.access_token){
		// login with access_token (from ingame)
		const access_token = await access_token_dao.find_by_username_and_token(req.body.name, req.body.access_token);

		if (!access_token){
			return res.status(401).json({
				message: "token not found"
			});
		}

		if (access_token.expires < Date.now()){
			return res.status(403).json({
				message: "token expired"
			});
		}

		// valid token and expiration time
		const token = createjwt(user, {
			audience: "minetest",
			expiresIn: parseInt((access_token.expires - Date.now()) / 1000)
		});
		res.send(token);

	} else {
		// invalid request
		return res.status(403).json({
			message: "Empty password/access_token not allowed"
		});
	}

});

File: web/token.go
package web

import (
	"blockexchange/core"
	"blockexchange/db"
	"blockexchange/types"
	"encoding/json"
	"net/http"
	"time"

	"github.com/sirupsen/logrus"
	"golang.org/x/crypto/bcrypt"
)

type TokenApi struct {
	AccessTokenRepo db.AccessTokenRepository
	UserRepo        db.UserRepository
}

func (api TokenApi) PostLogin(w http.ResponseWriter, r *http.Request) {
	login := types.Login{}
	err := json.NewDecoder(r.Body).Decode(&login)
	if err != nil {
		SendError(w, "decode: "+err.Error())
		return
	}

	logrus.WithFields(logrus.Fields{
		"Username": login.Username,
		"Password": login.Password,
		"Token":    login.Token,
	}).Debug("POST /api/token")

	user, err := api.UserRepo.GetUserByName(login.Username)
	if err != nil {
		SendError(w, "user: "+err.Error())
		return
	}
	if user == nil {
		SendError(w, "User not found")
		return
	}

	if login.Password != "" {
		// login with username / password
		err = bcrypt.CompareHashAndPassword([]byte(user.Hash), []byte(login.Password))
		if err != nil {
			SendError(w, err.Error())
			return
		}

		permissions := []types.JWTPermission{
			types.JWTPermissionUpload,
			types.JWTPermissionOverwrite,
			types.JWTPermissionManagement,
		}

		exp := time.Now().Unix() + (3600 * 24 * 180)
		token, err := core.CreateJWT(user, permissions, exp)
		if err != nil {
			SendError(w, err.Error())
			return
		}

		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(token))

	} else if login.Token != "" {
		// login with token
		access_token, err := api.AccessTokenRepo.GetAccessTokenByTokenAndUserID(login.Token, user.ID)
		if err != nil {
			SendError(w, err.Error())
			return
		}

		if access_token == nil {
			SendError(w, "access token not found")
			return
		}

		if access_token.Expires < (time.Now().Unix() * 1000) {
			SendError(w, "token expired")
			return
		}

		permissions := []types.JWTPermission{
			types.JWTPermissionUpload,
			types.JWTPermissionOverwrite,
		}
		token, err := core.CreateJWT(user, permissions, int64(access_token.Expires/1000))
		if err != nil {
			SendError(w, err.Error())
			return
		}
		api.AccessTokenRepo.IncrementAccessTokenUseCount(access_token.ID)

		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(token))

	} else {
		SendError(w, "Empty password/access_token not allowed")
	}

}
